src/tribler/core/components/popularity/community/popularity_community.py

```diff
@@ -140,3 +232,37 @@ def get_random_torrents(self) -> List[HealthInfo]:
 
         random_torrents = random.sample(checked_and_alive, num_torrents_to_send)
         return random_torrents
+
+    def _create_rendezvous_request(self) -> RendezvousRequestPayload:
+        challenge = RendezvousChallenge.create()
+        payload = RendezvousRequestPayload(challenge)
+        return payload
+
+    def _create_rendezvous_response(self, challenge: RendezvousChallenge) -> RendezvousResponsePayload:
+        signature = challenge.sign(self.my_peer.key)
+        payload = RendezvousResponsePayload(challenge, RendezvousSignature(signature))
+        return payload
+
+    def _handle_rendezvous_response(self, peer, raw_payload: RawRendezvousResponsePayload):
+        signature, _ = self.serializer.unpack_serializable(RendezvousSignature, raw_payload.signature)
+        challenge, _ = self.serializer.unpack_serializable(RendezvousChallenge, raw_payload.challenge)
+
+        expected_nonce = self.rendezvous_cache.get_rendezvous_challenge(peer) or EMPTY_PEER_CHALLENGE
+        if expected_nonce == EMPTY_PEER_CHALLENGE or expected_nonce != challenge.nonce:
+            self.logger.warning(f"Received invalid rendezvous response from {peer.mid}")
+            return
+
+        if not self.crypto.is_valid_signature(peer.key, raw_payload.challenge, signature.signature):
+            self.logger.warning(f"Received invalid signature from {peer.mid}")
+            return
+        else:
+            # This nonce has been burned.
+            self.rendezvous_cache.clear_peer_challenge(peer)
+
+        self.logger.debug(f"Received valid rendezvous response from {peer.mid}")
+        with db_session:
+            certificate = self.rdb.Certificate.get(public_key=peer.mid)
```

Comment:
To avoid possible database locking, it is better to either use here `db_session(immediate=True)` or `Certificate.get_for_update(public_key=peer.mid)` (the result is the same)